compiler/parser/RdrHsSyn.hs

```diff
@@ -2455,247 +2505,199 @@ mkTemplateTypeDecl
+  in [baseInstance, templateInstance]
+
+mkChoiceInstanceDecl :: String -> ChoiceData -> LHsDecl GhcPs
+mkChoiceInstanceDecl templateName ChoiceData{..} =
+  let mkTypeName = noLoc . HsTyVar NoExt NotPromoted . noLoc
+      mkAppTy :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+      mkAppTy ty1 ty2 = noLoc $ HsAppTy noExt ty1 ty2
+      choiceType = noLoc $ HsTyVar noExt NotPromoted cdChoiceName
+      returnType = noLoc $ HsParTy noExt cdChoiceReturnTy
+      choiceClass = mkTypeName $ qualifyDesugar $ mkClsOcc "Choice"
+      templateType = mkTypeName $ mkRdrUnqual $ mkTcOcc templateName
+      instanceType = choiceClass `mkAppTy` templateType `mkAppTy` choiceType `mkAppTy` returnType
+      mkVar = noLoc . HsVar noExt . noLoc . mkRdrUnqual . mkVarOcc
+      exerciseMethodBody =
+        mkVar $ prefixTemplateClassMethod $
+          "exercise" ++ templateName ++ occNameString (rdrNameOcc (unLoc cdChoiceName))
+      exerciseMethod = mkTemplateClassMethod "exercise" [] exerciseMethodBody Nothing
+  in instDecl $ classInstDecl (unLoc instanceType) $ listToBag [exerciseMethod]
+
+mkKeyInstanceDecl :: String -> LHsType GhcPs -> LHsDecl GhcPs
+mkKeyInstanceDecl templateName keyType =
+  let mkTypeName = noLoc . HsTyVar NoExt NotPromoted . noLoc
+      mkAppTy :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+      mkAppTy ty1 ty2 = noLoc $ HsAppTy noExt ty1 ty2
+      templateKeyClass = mkTypeName $ qualifyDesugar $ mkClsOcc "TemplateKey"
+      templateType = mkTypeName $ mkRdrUnqual $ mkTcOcc templateName
+      instanceType = templateKeyClass `mkAppTy` templateType `mkAppTy` keyType
+      mkVar = noLoc . HsVar noExt . noLoc . mkRdrUnqual . mkVarOcc
+      mkMethod methodName =
+        let methodBody = mkVar $ prefixTemplateClassMethod $ methodName ++ templateName in
+        mkTemplateClassMethod methodName [] methodBody Nothing
+      methods = map mkMethod ["key", "fetchByKey", "lookupByKey"]
+  in instDecl $ classInstDecl (unLoc instanceType) (listToBag methods)
+
+-- | Contruct a @data S = S {...}@ for a single choice 'S'.
 mkTemplateChoiceDecls
-  dataName conName controllers (L _ (choice@ChoiceData{..})) argPatType binds = do
-{
+  :: ChoiceData          -- choice 'S'
+  -> P [LHsDecl GhcPs]   -- resulting declarations
+mkTemplateChoiceDecls ChoiceData{..} = do
   -- Calculate data constructor info from the choice name and (maybe)
   -- record type.
-    let lname@(L nloc _name) = cdChoiceName
-        choiceName = L nloc (HsTyVar noExt NotPromoted lname)
-  ; choiceConInfo@(choiceConName, _, _) <-
-      case cdChoiceFields of
-        Nothing -> splitCon [choiceName]
-        Just fields -> splitCon [fields, choiceName]
-  ; dataDecl <- mkTemplateTypeDecl
-                  (combineLocs cdChoiceName
+  let lname@(L nloc _name) = cdChoiceName
+      choiceName = L nloc (HsTyVar noExt NotPromoted lname)
+  choiceConInfo <- splitCon $ maybe id (:) cdChoiceFields [choiceName]
```

Comment:
It would be nice to remove `splitCon`. We copied it from an older version of the compiler. We need to understand what replaced it.

Reply:
I think that's a good idea. I will probably do that in a follow-up PR though.

Reply:
Thanks. Yes, it's overdue!